packages/sources/por-indexer/src/endpoint/balance.ts

```diff
@@ -0,0 +1,108 @@
+import { InputParameters } from '@chainlink/external-adapter-framework/validation'
+import { PoRTotalBalanceEndpoint } from '@chainlink/external-adapter-framework/adapter/por'
+import { config } from '../config'
+import { balanceTransport } from '../transport/balance'
+import { AdapterRequest } from '@chainlink/external-adapter-framework/util'
+import { AdapterInputError } from '@chainlink/external-adapter-framework/validation/error'
+
+export const inputParameters = new InputParameters({
+  addresses: {
+    aliases: ['result'],
+    array: true,
+    type: {
+      address: {
+        type: 'string',
+        description: 'an address to get the balance of',
+        required: true,
+      },
+      network: {
+        description: 'The name of the target network protocol',
+        required: true,
+        type: 'string',
+      },
+      chainId: {
+        description: 'The name of the target chain',
+        required: true,
+        type: 'string',
+      },
+    },
+    description:
+      'An array of addresses to get the balances of (as an object with string `address` as an attribute)',
+    required: true,
+  },
+  minConfirmations: {
+    type: 'number',
+    default: 0,
+    description:
+      'Number of blocks that must have been confirmed after the point against which the balance is checked.',
+  },
+})
+
+export type BaseEndpointTypes = {
+  Parameters: typeof inputParameters.definition
+  Response: {
+    Result: string
+    Data: {
+      result: string
+    }
+  }
+  Settings: typeof config.settings
+}
+
+export const endpoint = new PoRTotalBalanceEndpoint({
+  name: 'balance',
+  aliases: ['index'],
+  transport: balanceTransport,
+  inputParameters,
+  customInputValidation: (
+    req: AdapterRequest<typeof inputParameters.validated>,
+    settings: typeof config.settings,
+  ): AdapterInputError | undefined => {
+    const addresses = req.requestContext.data.addresses
+    if (addresses.length === 0) {
+      throw new AdapterInputError({
+        statusCode: 400,
+        message: `Input, at 'addresses' or 'result' path, must be a non-empty array.`,
+      })
+    }
+
+    const btcMainnet = addresses.some(
+      (a) => `${a.network}_${a.chainId}`.toUpperCase() === 'BITCOIN_MAINNET',
+    )
+    const btcTestnet = addresses.some(
+      (a) => `${a.network}_${a.chainId}`.toUpperCase() === 'BITCOIN_TESTNET',
+    )
+    const dogeMainnet = addresses.some(
+      (a) => `${a.network}_${a.chainId}`.toUpperCase() === 'DOGECOIN_MAINNET',
+    )
+    const dogeTestnet = addresses.some(
+      (a) => `${a.network}_${a.chainId}`.toUpperCase() === 'DOGECOIN_TESTNET',
+    )
```

Comment:
I think in reality most of these requests if not all of them just have one network/chainId per request. So 3 of these loops would be going through the address set without finding anything. Could we change this to find all of the unique network/chainId combos with just a single pass through the address set? I think address sets can get pretty large so it might make a difference especially since this runs on every request.